### ISSUE TITLE
D1b-3 · Held-out eval harness + decision gate

### ISSUE THREAD

**Issue by bouthilx** [bouthilx/paperext#53]
Part of #44 (D1b umbrella). **Does the agent earn the right to run across D1c–D1e?** Built and run on **models only**, before any rollout.

## Why the obvious eval is invalid

**1 · This is not a classification problem.** The classes are not mutually exclusive — they *nest*. `transformer` contains `vision transformer`, so placing a name one level up from the reference is usually defensible, while a flat metric scores it identically to putting it under `graph neural network`. They are also highly skewed: at the milabench cut the models pool is 45.4% `Other`, then `transformer`, then six buckets in the single digits, so "always answer `Other`" already scores ~0.45 and an accuracy figure mostly measures the prior.

**2 · The reference is not a hand-built tree — it is an LLM-in-the-loop artifact.** `ontology_mapping/build_models_tree.py:56,385`: candidates came from `Levenshtein.distance`, the top-5 were shown to a human who typed a digit, and `h` called **gpt-4o** for a suggestion that was *prepended to the proposition list*. So agreement with `data/categorized_models.json` partly measures **reproducing a 2024 gpt-4o + Levenshtein pipeline**, including its first-item anchoring bias.

**3 · The repo is public.** `data/categorized_models.json` may sit in the pinned model's training data, so raw agreement is inflatable by recall rather than judgment.

Together these rule out "accuracy against the reference" as the gate. They do **not** rule out measuring the reference — they change what the measurement means.

## The primary metric: blind pairwise win rate

Per gate item, render the reference's placement and the agent's placement as **anonymized ancestor paths**, in randomized A/B order, alongside the paper evidence, and ask a blind judge which is the better placement:

```
s_i = 0.5  if the two placements agree           (a tie by definition)
s_i = 1.0  if the judge prefers the agent
s_i = 0.5  if the judge says equally defensible
s_i = 0.0  if the judge prefers the reference

W = mean(s_i) = 0.5 + 0.5·(g − l)
```

Why this and not agreement:

- **The ceiling is 0.5 analytically, not empirically.** An equally good annotator wins half the blind comparisons. Nothing has to be estimated, so the gate carries no ceiling-estimation error. (A ceiling measured from ~50–60 blind re-mappings is only pinned to ±9–14pp; gating relative to it would be a threshold with a ±14pp denominator, which is not a decision rule.)
- **It degrades gracefully under memorization.** If the agent recalls the published tree it produces agreements, which score 0.5 — pushing W *toward* the ceiling, never above it. Raw agreement is inflated by exactly the same behaviour.
- **"The legacy tree is wrong" becomes a first-class outcome** (`g`), cleanly separated from "the agent is wrong" (`l`) and "defensibly different" (tie).
- **Adjudication cost scales with disagreements, not items** — agreements need no attention at all.

Report `l`, `t`, `g` separately. `g` — the rate at which the agent improves on the legacy tree — is independently useful input to the D1c cleanup pass.

## Reported alongside: the shape of the result

W is the gate; these are what make the result legible and protect the report:

**A · Aggregate shape.** Per-category **share**, agent vs reference, at the milabench cut and at depths 1–2 — "12.4% of names under `transformer` vs 11.8% in the reference" — reported cumulatively at each level, since ancestor counts are cumulative by construction. Plus **rank agreement** of the category ordering (Spearman ρ, top-5 set overlap) and **total-variation distance** between the normalized count vectors.

**B · Hierarchy-aware per-item agreement.** Compare *ancestor sets*, not labels: **hierarchical precision/recall/F1** (`hP = |anc(pred) ∩ anc(ref)| / |anc(pred)|`, `hR` likewise over `|anc(ref)|`) — placing at `transformer` when the reference says `transformer > vision transformer` scores full precision, partial recall; **same-lineage rate** (prediction is an ancestor or descendant of the reference); **LCA depth** for everything off-lineage.

**C · Coverage — reference-free, and the actual goal of D1.** Distinct-name coverage (1153/2342 mapped today, 49%); **mention-weighted coverage**, which is what actually moves rankings; shrinkage of the `Other` bucket.

**Baselines, reported against every agreement number**: the majority class (~0.45 at cut), and a **Levenshtein-nearest-neighbour placer** — free to compute, and *the exact mechanism that generated the reference*. If a pinned frontier model does not beat it by a wide margin, the high cost band is not earning its keep.

## Reference-free probes (no subjectivity at all)

- **No-op control** — 100 already-mapped names, **not** ablated. Correct behaviour is known with zero subjectivity: a no-op, or at most `update_description`. Target no-op rate ≥ 0.95, and **zero destructive ops** on nodes outside the candidate set.
- **Policy conformance** — from the corpus alias pairs: ~73 suffix-qualified pairs (`ResNet`→`ResNet-50`, `RoBERTa`→`RoBERTa-large`) **must become child nodes**, ~30 true spelling variants (`ResNet18`/`ResNet-18`) **must become surfaces**. The locked granularity policy fixes the answer without any reference tree. Zero human effort.
- **Ambiguity 2×2** — natural collisions are too few (4 in models), so construct: ~40 acronym items with grounding *stripped* (correct = abstain + `review_note`), the same 40 **with** grounding (correct = resolve), and ~20 injected homonyms. Score **Youden's J = recall_abstain + (1 − false_abstain) − 1** — abstention recall alone is gameable by always abstaining, which is why the matched controls are mandatory.
- **Self-consistency** — k=3 runs, fresh ablation each time, identical payload verified by hash. Report triple-agreement and Fleiss' κ. This **upper-bounds** achievable agreement with any reference, and doubles as a validity check: if agreement-with-legacy *exceeds* self-consistency, there is a leak or memorization — investigate before believing anything.
- **Memorization canary** — before anything else, probe the pinned model out-of-band for the top-level structure of the published `categorized_models.json`. Five minutes, decisive. If it fires, the raw-agreement floor below is void and only W and the reference-free criteria count. Run it and record the result here.
- **Schema / apply / invariant validity**, and **cost/latency** per decision plus the projected rollout total.

## Sample

Pool = already-mapped **leaf** names (non-leaf targets are excluded — ablating them orphans a subtree). Stratify on the two axes that actually predict difficulty: **candidate-anchoring** (`parent` / `some` / `cold` — see the retrieval numbers in D1b-2) and **cut class**, so all report-facing buckets get support. Splits sealed in **D1b-1**:

| Split | N | Purpose |
|---|---|---|
| **dev** | 200 | prompt iteration, retrieval tuning, unlimited re-runs |
| **gate** | 300 | sealed; opened at most twice; the gate number |
| **reserve** | rest | future re-gates (C1 re-run, DAG activation, model swap) |

**Sizing.** With plausible values (agreement 0.50; on the disagreements l=0.15, t=0.25, g=0.10) → W ≈ 0.475, SD(s) ≈ 0.25. At N=300, SE ≈ 0.014 and power to detect W=0.50 against `H₀: W ≤ 0.45` is ~97%. **At N=100 power is only ~64% — a 100-item gate is not worth running.**

**Protocol: leave-one-out, non-accumulating.** Each item is decided against a fresh single-item-ablated copy, so items are i.i.d. and the CIs are honest. Ablating all 300 at once would destroy sibling context and correlate the errors. A separate **sequential-replay diagnostic** (one accumulating pass, reported not gated) is the only thing that predicts what a real D1c run does: chaining rate, end-of-run invariants, node-count drift, duplicate-concept rate.

## The gate

Pre-registered — these numbers go in this issue **before** the gate run. Pass requires all six.

1. **Primary.** One-sided 95% lower bound on `W` ≥ **0.45** on the sealed 300 (bootstrap over items, fixed seed) — i.e. we can rule out that the agent is more than 5 points worse than the reference under blind pairwise judgment. Headline reported as **W@coverage=0.85** with the full risk–coverage curve, so abstention cannot inflate it.
2. **Report-facing floor** *(void if the memorization canary fires)*. Cut-level agreement ≥ max(0.70, Levenshtein baseline + 0.10), κ ≥ 0.55, **and macro-recall over the non-`Other` cut buckets ≥ 0.60** — the macro-recall clause is what blocks the always-`Other` degenerate that the 45% prevalence would otherwise reward. Plus Spearman ρ ≥ 0.9 on the category ranking and top-5 unchanged.
3. **Harmlessness.** Schema-valid 100%; apply-success ≥ 99%; invariants clean 100%; no-op control ≥ 0.95 with zero destructive ops outside the candidate set.
4. **Consistency.** k=3 triple-agreement ≥ 0.75, and self-consistency > agreement-with-legacy.
5. **Abstention & policy.** Youden's J ≥ 0.70; coverage ≥ 0.85; policy conformance ≥ 0.90.
6. **Ignore detection** — its own metric, because it is much less subjective and it silently deletes data: `mark_ignore` precision ≥ 0.80, recall ≥ 0.60. Precision matters more; a false ignore removes a real entity from every downstream count.

**Coverage must also rise** — mention-weighted coverage materially above today's 49%, with `Other` correspondingly smaller. And **any category whose share moves by more than 25% needs a written, adjudicated justification**: the legacy tree is subjective and sometimes wrong, so a large move is *allowed*, it just has to be explained rather than silently absorbed. If the deviations are mostly `legacy-wrong`, that is a signal to refresh the reference and a finding for the D1c cleanup phase — not a reason to fail the agent.

## Human effort — bounded

Adjudication is done by a **different-vendor** frontier model (agent Claude → judge OpenAI, or the reverse), blind and order-randomized. The human **audits the judge**, rather than doing the adjudication.

| Task | Items | Time |
|---|---|---|
| Rubric calibration pilot | 20 | 30 min |
| Blind re-map (descriptive ceiling) | 60 | 90 min |
| Audit of LLM adjudications | 40 | 30 min |
| **Total, one-time** | **~120** | **≈ 2.5 h** |
| Per later agent revision | ~15 changed pairs | ~15 min |

A **verdict cache** keyed by `sha256(item_id ‖ reference_placement ‖ agent_placement)` keeps revisions cheap — a prompt tweak that changes 40 of 300 decisions costs 40 adjudications, not 300. The 60 blind re-mappings are a permanent asset, reusable against every future agent version.

**Guardrails, without which the W number is unfounded**: judge order-swap consistency ≥ 0.80 on duplicated pairs; on the 40-item human audit the judge's pro-agent rate must not exceed the human's by more than ~8 points, else fall back to full human adjudication of the gate set (~2 h) and say so; and the judge must be a different vendor from the agent, since a same-family judge preferring its own output would bias W upward — the one direction that matters.

## Deliverables

`categorize/evaluate.py` (LOO harness, k-run repeats, control sets, sequential-replay diagnostic), metrics, blind adjudication + verdict cache, the probes, a report generator, and a `categorize-eval` CLI. Data under `data/ontology/eval/models/`. Committed report: the numbers, the adjudication labels, and the gate verdict.

## Acceptance (zero API calls)

Metric functions checked against synthetic cases with hand-computed κ / macro-recall / W and CI; bootstrap reproducible under a fixed seed; verdict-cache keys stable across reorderings; adjudication rendering provably order-blind (swapping A/B yields a mirror record, not a different item); and the whole harness runs end-to-end against a **stub agent replaying a recorded `decisions.jsonl`**, producing a full report with no network.

## Scope of the licence a pass grants

D1c (models), D1d-datasets and D1e (research fields), each with a **120-item mini-gate** (`LB₉₅(W) ≥ 0.42`), because the tree shapes differ sharply — datasets is 18 roots / 502 nodes (a much wider placement choice), domains 3 roots / 1536 nodes (very deep).

It does **not** license **D1d-libraries**. There is no libraries tree, so a held-out *re-mapping* eval cannot measure the task at all — building a tree from 738 names is a **construction** problem needing its own gate (coverage, duplicate-concept rate, depth balance, adjudicated pairwise against a small hand-built seed). Flagged as separately gated.

## Out of scope

The action vocabulary, applier, ablation and splits → D1b-1; prompt and agent loop → D1b-2. Running the agent across dimensions → D1c–D1e.
